cumulus/client/cirrus-executor/src/bundle_processor.rs

```diff
@@ -285,64 +282,62 @@ where
-						"Generating the signed execution receipt for #{}",
-						header_hash
-					);
+		let best_execution_chain_number =
+			<NumberFor<Block>>::decode(&mut best_execution_chain_number.encode().as_slice())
+				.expect("Primary number and secondary number must use the same type; qed");
 
-					let signed_execution_receipt = SignedExecutionReceipt {
-						execution_receipt,
-						signature: ExecutorSignature::decode(&mut signature.as_slice()).map_err(
-							|err| {
-								sp_blockchain::Error::Application(Box::from(format!(
-									"Failed to decode the signature of execution receipt: {err}"
-								)))
-							},
-						)?,
-						signer: executor_id,
-					};
-
-					if let Err(e) = self
-						.execution_receipt_sender
-						.unbounded_send(signed_execution_receipt.clone())
-					{
-						tracing::error!(target: LOG_TARGET, error = ?e, "Failed to send signed execution receipt");
-					}
+		assert!(
+			header_number > best_execution_chain_number,
+			"Consensus chain number must larger than execution chain number by at least 1"
+		);
 
-					// Return `Some(_)` to broadcast ER to all farmers via unsigned extrinsic.
-					Ok(Some(signed_execution_receipt))
-				},
-				Ok(None) => Err(sp_blockchain::Error::Application(Box::from(
-					"This should not happen as the existence of key was just checked",
-				))),
-				Err(error) => Err(sp_blockchain::Error::Application(Box::from(format!(
-					"Error occurred when signing the execution receipt: {error}"
-				)))),
-			}
+		// Ideally, the receipt of current block will be included in the next block, i.e., no
+		// missing receipts.
+		if header_number == best_execution_chain_number + One::one() {
+			self.try_sign_and_send_receipt(primary_hash, execution_receipt)
 		} else {
-			Ok(None)
+			// Receipts for some previous blocks are missing.
+			let max_drift = self
+				.primary_chain_client
+				.runtime_api()
+				.maximum_receipt_drift(&BlockId::Hash(primary_hash))?;
+
+			let max_drift = <NumberFor<Block>>::decode(&mut max_drift.encode().as_slice())
+				.expect("Primary number and secondary number must use the same type; qed");
+
+			let max_allowed = (best_execution_chain_number + max_drift).min(header_number);
+
+			// TODO: parallelize and avoid spamming the missing receipts?
+			let mut to_send = best_execution_chain_number;
+			while to_send <= max_allowed {
+				let block_hash = self.client.hash(to_send)?.ok_or_else(|| {
+					sp_blockchain::Error::Backend(format!("Hash for Block {:?} not found", to_send))
+				})?;
+
+				match crate::aux_schema::load_execution_receipt(&*self.client, block_hash)? {
+					Some(receipt) => {
+						self.try_sign_and_send_receipt(primary_hash, receipt)?;
+					},
+					None => {
+						// TODO: Re-generate the receipt if it has been pruned?
+						// But what if the block body is also pruned? That being said, we have to
+						// fetch the block body from somewhere(maybe DSN) to rebuild the block?
+						todo!("Re-generate the receipt?");
```

Comment:
The executor also only keeps a number of the receipts locally, the older receipts will be pruned. If the receipt is greatly lagging behind, when the major sync is done, the receipts which are not yet submitted to the primary chain but have been pruned on the executor side, given that the primary node is a full node, the older block body is also removed, the receipt can not be regenerated using the typical way. Thoughts? @nazar-pc

Reply:
Okay, this problem will be solved by [Post-MVP](https://www.notion.so/subspacelabs/Decoupled-Execution-Research-Liu-Cheng-84c621badb3b47939aaacee1e554a137#3e67edf878be442e84a6c77d467a6309), will update the TODO.

Reply:
Updated, please have a look at this note https://github.com/subspace/subspace/pull/446/commits/f18e73b1d8649c241eff5958ae0c2a8dcb113703#diff-6081dc0064f80697e0c8826a8774281dc237e4b4a0ecbf6ae856f6907d71b4daR336-R337.